File: source/Knyaz.Optimus.Tests/UnitTests/Dom/HtmlImageElementTests.cs
using System.Threading;
using System.Threading.Tasks;
using Knyaz.Optimus.Dom;
using Knyaz.Optimus.Dom.Elements;
using Knyaz.Optimus.Tools;
using Moq;
using NUnit.Framework;

namespace Knyaz.Optimus.Tests.Dom
{
	[TestFixture]
	public class HtmlImageElementTests
	{
		private Document _document;

		[SetUp]
		public void SetUp()
		{
			_document = DomImplementation.Instance.CreateHtmlDocument();
		}
		
		[Test]
		public void ImageOnError()
		{
			_document.GetImage = s => Task.Run(() => (IImage)null);
			
			var img = (HtmlImageElement)_document.CreateElement(TagsNames.Img);
			var errorSignal = new ManualResetEvent(false);
			img.OnError += evt => { errorSignal.Set(); };

			img.Src = "http://noimage.jpg";

			Assert.IsTrue(errorSignal.WaitOne(1000));
		}
		
		[Test]
		public void ImageOnErrorTwice()
		{
			_document.GetImage = s => Task.Run(() => (IImage)null);
			var img = (HtmlImageElement)_document.CreateElement(TagsNames.Img);
			var errorSignal = new ManualResetEvent(false);
			img.OnError += evt => { errorSignal.Set(); };

			img.Src = "http://noimage.jpg";

			Assert.IsTrue(errorSignal.WaitOne(1000), "first error");

			errorSignal.Reset();
			img.Src = "http://noimage.jpg";
			
			Assert.IsTrue(errorSignal.WaitOne(1000), "second error");
		}

		[Test]
		public void ImageOnLoad()
		{
			var imageData = Mock.Of<IImage>(x => x.Width == 8 && x.Height == 4);
			_document.GetImage = s => Task.Run(() => imageData);

			var img = (HtmlImageElement)_document.CreateElement("img");
			var loadSignal = new ManualResetEvent(false);
			img.OnLoad += evt => { loadSignal.Set(); };

			img.Src = "http://image.jpg";

			Assert.IsTrue(loadSignal.WaitOne(1000));
			Assert.AreEqual(8, img.NaturalWidth);
			Assert.AreEqual(4, img.NaturalHeight);
		}
	}
}

File: source/Knyaz.Optimus.Tests/Resources/JsTests/HtmlImageElementTests.js
Test("HtmlImageElementTests", {
    "NoArgumentsCtor": {
        run: function () {
            var img = new Image();
            Assert.AreEqual(0, img.width, "width");
            Assert.AreEqual(0, img.height, "height");
            Assert.AreEqual(true, img.complete, "complete");
        }
    },
    "OneArgumentCtor":{
        run:function () {
            var img = new Image(100);
            Assert.AreEqual(100, img.width, "width");
            Assert.AreEqual(0, img.height, "height");
            Assert.AreEqual(true, img.complete, "complete");
        }
    },
    "TwoArgumentsCtor": {
        run: function () {
            var img = new Image(100,200);
            Assert.AreEqual(100, img.width, "width");
            Assert.AreEqual(200, img.height, "height");
            Assert.AreEqual(100, img.getAttribute("width"), "attribute 'width'");
            Assert.AreEqual(200, img.getAttribute("height"), "attribute 'height'");
            Assert.AreEqual(true, img.complete, "complete");
        }
    },
    "SetWidthAndHeight":{
        run: function () {
            var img = new Image(100,200);
            img.width = 300;
            img.height = 400;
            Assert.AreEqual(300, img.width, "width");
            Assert.AreEqual(400, img.height, "height");
        }
    },
    "TypeOfNewImage":{
        run: function(){
            var img = new Image(10,10);
            Assert.AreEqual("[object HTMLImageElement]", img.toString());
        }
    },
    "PrototypeOfNewImage":{
        run: function(){
            var img = new Image(10,10);
            Assert.AreEqual("[object HTMLImageElementPrototype]", Object.getPrototypeOf(img).toString());
        }
    },
    "NewImageInstanceOfHTMLImageElement":{
        run: function(){
            Assert.AreEqual(true, new Image(1,1) instanceof HTMLImageElement);
        }
    }
});